stdlib/runtime/string.gr

```diff
@@ -37,6 +33,98 @@ primitive (||) : (Bool, Bool) -> Bool = "@or"
 
 enum StringList { [], [...](String, StringList) }
 
+let _RUNTIME_TYPE_METADATA_PTR = 256n
+
+@disableGC
+let findTypeMetadata = (moduleId, typeId) => {
+  let mut metadataPtr = WasmI32.load(_RUNTIME_TYPE_METADATA_PTR, 0n)
+  let mut modData = -1n
+  let mut modTypesCount = 0n
+  while (metadataPtr != 0n) {
+    if (WasmI32.load(metadataPtr, 4n) == moduleId) {
+      modData = metadataPtr + 12n
+      modTypesCount = WasmI32.load(metadataPtr, 8n)
+      break
+    }
+    metadataPtr = WasmI32.load(metadataPtr, 0n)
+  }
+
+  if (modData == -1n) {
+    -1n
+  } else {
+    let mut typeData = -1n
+    for (let mut i = 0n; i < modTypesCount; i += 1n) {
+      if (WasmI32.load(modData, 4n) == typeId) {
+        typeData = modData
+        break
+      }
+      modData += WasmI32.load(modData, 0n)
+    }
+
+    typeData
+  }
+}
+
+@disableGC
+let getVariantName = (variant) => {
+  let moduleId = WasmI32.load(variant, 4n) >> 1n
+  let typeId = WasmI32.load(variant, 8n) >> 1n
+  let variantId = WasmI32.load(variant, 12n) >> 1n
+
+  let mut block = findTypeMetadata(moduleId, typeId)
+
+  if (block == -1n) {
+    -1n
+  } else {
+    let sectionLength = WasmI32.load(block, 0n)
+    block += 8n
+
+    let end = block + sectionLength
+    let mut result = -1n
+    while (block < end) {
+      if (WasmI32.load(block, 4n) == variantId) {
+        let length = WasmI32.load(block, 8n)
+        let str = allocateString(length)
+        Memory.copy(str + 8n, block + 12n, length)
+        result = str
+        break
+      }
+      block += WasmI32.load(block, 0n)
+    }
+
+    result
+  }
+}
+
+@disableGC
+let getRecordFieldNames = (record_) => {
+  let moduleId = WasmI32.load(record_, 4n) >> 1n
+  let typeId = WasmI32.load(record_, 8n) >> 1n
+  let arity = WasmI32.load(record_, 12n)
+
+  let mut fields = findTypeMetadata(moduleId, typeId)
+
+  if (fields == -1n) {
+    -1n
+  } else {
+    fields += 8n
+    
+    let fieldArray = allocateArray(arity)
+
+    let mut fieldOffset = 0n
+    for (let mut i = 0n; i < arity; i += 1n) {
+      let fieldLength = WasmI32.load(fields + fieldOffset, 4n)
+      let fieldName = allocateString(fieldLength)
+      Memory.copy(fieldName + 8n, fields + fieldOffset + 8n, fieldLength)
+      WasmI32.store(fieldArray + (i * 4n), fieldName, 8n)
+
+      fieldOffset += WasmI32.load(fields + fieldOffset, 0n)
+    }
+
+    fieldArray
+  }
+}
```

Comment:
Are there opportunities for constants with a bunch of these hardcoded offsets and stuff? I forget when we can and can't use them.

Reply:
We can't use them here, though I realized it actually wouldn't be that hard to support it. https://github.com/grain-lang/grain/issues/571